doc/sonic-build-system/SONiC-OpenSSL-FIPS-140-3.md

```diff
@@ -0,0 +1,140 @@
+
+The SymCrypt Engine is one of the implementation to support FIPS, The [wolfSSL engine](https://github.com/wolfSSL/wolfEngine) is another option.
+
+### OpenSSL configuration for SymCrypt Engine
+
+/usr/lib/ssl/openssl-fips.cnf 
+```
+openssl_conf = openssl_init
+[ openssl_init ]
+engines = engine_section
+
+[ engine_section ]
+symcrypt = symcrypt_section
+
+[ symcrypt_section ]
+engine_id = symcrypt
+dynamic_path = /usr/lib/x86_64-linux-gnu/libsymcryptengine.so
+default_algorithms = ALL
+```
+
+### OpenSSL configuration enhancement
+When fips=1 is set in /proc/cmdline, the OpenSSL default config file is changed to "/usr/lib/ssl/openssl-fips.cnf".
+
+### SymCrypt OpenSSL Engine debian package
+Provide SymCrypt OpenSSL debian package.
+Package name: symcrypt-openssl
+Current version: 0.1
+
+Package file name example: symcrypt-openssl_0.1_amd64.deb
+Files in the packages:
+```
+/usr/lib/ssl/openssl.cnf
+/usr/lib/ssl/openssl-fips.cnf
+/usr/lib/x86_64-linux-gnu/libsymcrypt.so
+/usr/lib/x86_64-linux-gnu/libsymcryptengine.so
+```
+
+## Kerberos Cryptographic Module
+Kerberos will use the builtin cryptographic module by default, but it allows to change the build option to use OpenSSl, see [MIT Kerberos features](https://web.mit.edu/kerberos/krb5-1.13/doc/mitK5features.html). SONiC will change the build option to use OpenSSL instead of the builtin one. It is not configurable to use the Kerberos builtin cryptographic module when OpenSSL used.
+
+## Golang Cryptographic Module
+Golang has its own cryptographic module (see [crypto](https://github.com/golang/go/tree/master/src/crypto)) without FIPS supports. There are some branches with branch name starting with "dev.boringcrypto" (see [golang branches](https://github.com/golang/go/branches/all?query=dev.boringcrypto)), changing the Golang cryptographic APIs' referenece to use [BoringSSL](https://github.com/google/boringssl). Although BoringSSL is an open source project, but it used by Google only, not intened for general use.
+
+To support FIPS for Golang, RedHat offers an alternative solution (see [here](https://developers.redhat.com/blog/2019/06/24/go-and-fips-140-2-on-red-hat-enterprise-linux)), it builds on top of the Golang's dev.bringcrypt branches, has ability to call into OpenSSL, not BoringSSL. SONiC can reuse the RedHat sulotion, one difference is that RedHat supports FIPS for OpenSSL directly, SONiC uses OpenSSL Engine.
+
+How OpenSSL Engine works in Golang?
+![Golang API!](images/golang-api.png)
```

Comment:
>golang-api.png [](http://example.com/codeflow?start=22&length=14)

The diamond in flow chart is weird.

Reply:
First one, GOLANG_FIPS=1 or fips=1
Second one, ENABLE_FIPS=1 or fips=1